### PR TITLE
InMemory escalated transactions throw exception via Rollback

### PR DESCRIPTION
**Pull request by WilliamBZA**
InMemoryTransaction storage with concurrent access could result in valid
Exceptions being thrown, but the exception was thrown on the wrong
thread since the transaction wasn't being rolled back via DTC correctly

Connects to #3244

### CODE REVIEW

File: src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+                    await unenlistedSession.CompleteAsync();
+
+                    tx.Complete();
+                }
+            });
+        }
+    }
+
+    public class DummyEnlistmentNotification : IEnlistmentNotification
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs]: Should we rename to something like `EnlistmentWhichEnforcesDtcEscalation`?

File: src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            };
+
+            var persister = new InMemorySagaPersister();
+            var storageAdapter = new InMemoryTransactionalSynchronizedStorageAdapter();
+            var insertSession = new InMemorySynchronizedStorageSession();
+
+            await persister.Save(saga, SagaMetadataHelper.GetMetadata<TestSaga>(saga), insertSession, new ContextBag());
+            await insertSession.CompleteAsync();
+
+            var expectedException = Assert.Throws<TransactionAbortedException>(async () =>
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs]: @hmemcpy Shouldn't we start using `Assert.That` in preparation for NUnit 3 when combined with `async`?

Comment (WilliamBZA) [inline src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs]: Yeah, we should. I'll change.

Comment (hmemcpy) [inline src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs]: @danielmarbach @WilliamBZA I'm pretty sure it ends up calling `Assert.That` under covers, but honestly I don't even remember the original problem in the first place (something to do with not being able to handle async void methods...?)

Comment (WilliamBZA) [inline src/NServiceBus.Core.Tests/Persistence/InMemory/When_persisting_a_saga_with_InMemory_and_an_escalated_DTC_transaction.cs]: Both seem to handle async anonymous delegates just fine :shrugs:

I personally *feel* that the syntax of Assert.That is just cleaner.

**Comment by danielmarbach** [comment]
Can you remove https://github.com/Particular/NServiceBus/blob/bugix-inmemorywithsagasconcurrency/src/NServiceBus.Core/Persistence/InMemory/InMemoryTransactionalSynchronizedStorageAdapter.cs#L37 the private modifier?

**Comment by danielmarbach** [comment]
@WilliamBZA Fixed a formatting issue in `InMemoryTransactionalSynchronizedStorageAdapter` and merged.

**Pull request merged**